Added locations for hidden chest locations, reworked options a bit.

ScipioWright opened the pull request:
Added an option for whether taking the main path, side paths, or the main world. Changed the boss and orb options to take this into account instead of having their own separate options. Might end up changing it back later, feeling it out still. Slight change to item rates.

Draft because I'm not done yet, just putting it up so feedback can be a thing if desired.

ScipioWright committed:
Added locations for hidden chests.
Added an option for whether taking the main path, side paths, or the main world. Changed the boss and orb options to take this into account instead of having their own separate options. Might end up changing it back later, feeling it out still.
Slight change to item rates.

heinermann commented on worlds/noita/Regions.py:
```diff
@@ -15,10 +15,20 @@ def create_region(world: MultiWorld, player: int, region_name: str) -> Region:
         ltype = location_data.ltype
         flag = location_data.flag
-        if flag == 0 or ltype == "orb" and flag <= opt_orbs or ltype == "boss" and flag <= opt_bosses:
+        if flag == 0 and ltype != "orb" or "boss":
```
```py
if flag == 0 and ltype != "orb" and ltype != "boss":
```

I'm assuming this is what you meant?

ScipioWright replied:
Oh, right, order of operations, duh. Can I do something like `if flag == 0 and ltype != ("orb" or "boss")`?

heinermann replied:
No, it's more that you're not telling it the full statement. You're writing it in a way that would be intuitive in english but it's a computer so it doesn't work like that. 

This is how the original would be evaluated:

1: `if flag == 0 and ltype != "orb" or "boss":`
2: `if True and ltype != "orb" or "boss":`
3: `if True and False or "boss":`
4: `if False or "boss":` (and evaluated first)
5: `if "boss":` 
6: `if True:` (strings are truthy)


The second one:

1: `if flag == 0 and ltype != ("orb" or "boss")`  (because `"orb"` is truthy it chooses that)
2: `if flag == 0 and ltype != "orb"`
3: `if True and ltype != "orb"`
4: `if True and False`

heinermann replied:
I think the alternative would be something like

```py
if flag == 0 and ltype not in ["orb", "boss"]:
```

ScipioWright committed:
Minor edit per comments

ScipioWright commented:
I'm thinking something like the below:

```
if ltype == "hc" and flag <= opt_hc:
	for i in range(checks_per_region):
		# something to tell it to just take the first X values, based on a slider in Options
		# maybe we need some sort of key stating each region's starting values, so we can say "for each key, add key + i for i in range(checks_per_region)
		# or alternatively make it only generate the specific number of values based on the options
```

Still thinking this through, just jotting down ideas at work to get it out of my head

ScipioWright committed:
Still working on this, just putting progress up for the night. Currently trying to ditch total_locations since we have hidden chests now (and will have pedestals soon enough). Currently it generates properly if you turn all the options to main world, choose 5 on the hidden chest slider, and choose 153 total locations.

ScipioWright committed:
Total Locations is out.
Hidden chests is in.
Pedestals is also in, but commented out until they're implemented in the mod.

ScipioWright commented:
Alright, figured out a lot today (while sick, no less). Hidden chest checks are ready to go as soon as I implement it properly in the mod. Leaving this as a draft until then.
Also included similar code for the pedestal checks that will come eventually. I commented those out for now, but they seem good to go once pedestals are implemented into the mod.

heinermann commented on worlds/noita/Locations.py:
```diff
@@ -118,71 +119,86 @@ class HiddenChests(IntEnum):
     },
     "Overgrown Cavern": {
         "Overgrown Cavern Chest": LocationData(112200, "hc", HiddenChests.main_world),
+        "Overgrown Cavern Pedestal": LocationData(112700, "peds", Pedestals.main_world),
```
We'll need ranges otherwise there is only 1 chest and 1 pedestal

ScipioWright replied:
I'm doing the numbered chests and pedestals in Regions instead -- it made the most sense to me to do it there, since we need to use the slider to determine the number of checks needed, and we don't currently have access to that value in Locations.

heinermann replied:
Ah ok I see.

ScipioWright committed:
Updated location_name_to_id to iterate the hidden chest and pedestal names, so that the names get properly sent to Noita in the DataPackage

heinermann commented on worlds/noita/Locations.py:
```diff
@@ -161,17 +219,12 @@ class Bosses(IntEnum):
 for location_group in location_region_mapping.values():
     for locname, locinfo in location_group.items():
-        location_name_to_id.update({locname: locinfo.id})
+        if locname.endswith("Chest") or locname.endswith("Pedestal"):
```
We should use the same mechanism here as we do in `Regions.py`, something like `if locitem.ltype in ["hc", "peds"]:`

Mainly to avoid cases later if we (or someone down the line) puts things in unique chests/pedestals. Comments indicating what each `ltype` is above `location_region_mapping` would help too.

ScipioWright replied:
Updated and tested

heinermann commented on worlds/noita/Regions.py:
```diff
@@ -15,27 +15,30 @@ def create_region(world: MultiWorld, player: int, region_name: str) -> Region:
+                location.progress_type = LocationProgressType.DEFAULT
+                new_region.locations.append(location)
+
+        # if ltype == "peds" and flag <= opt_paths:
```
I'm guessing this is commented out until pedestals are implemented?

ScipioWright replied:
That's correct, just wrote it in advance

ScipioWright committed:
Minor update per Heinermann's comments
